### PR TITLE
1494 add grpc gateway compiler rules

### PR DESCRIPTION
**Pull request by Mistobaan**
See #1494

### CODE REVIEW

File: go/private/repositories.bzl [hunk 1]
```diff
@@ -103,6 +104,9 @@ def go_rules_dependencies():
       name = "gogo_special_proto",
   )
 
+  _maybe(googleapis_repository,
```
Comment (jayconrod) [inline go/private/repositories.bzl]: It doesn't seem like this needs to be a new repository rule. Can you use `git_repository` with `overlay` instead?

File: go/private/repositories.bzl [hunk 2]
```diff
@@ -134,6 +138,14 @@ def go_rules_dependencies():
       # importpath = "google.golang.org/genproto",
   )
 
+  # GRPC Gateway
+  _maybe(git_repository,
+      name = "com_github_grpc_ecosystem_grpc_gateway",
+      remote = "https://github.com/Mistobaan/grpc-gateway", # FIXME once 266 is merged
```
Comment (jayconrod) [inline go/private/repositories.bzl]: What is 266? The issue with that number in `github.com/grpc-ecosystem/grpc-gateway` is unrelated.

This should be merged before proceeding with this PR. We should depend on the real repo.

File: proto/BUILD.bazel [hunk 3]
```diff
@@ -40,6 +40,50 @@ go_proto_compiler(
...
+    deps = [
+        "@com_github_golang_protobuf//proto:go_default_library",
+        "@com_github_grpc_ecosystem_grpc_gateway//runtime:go_default_library",
+        "@com_github_grpc_ecosystem_grpc_gateway//utilities:go_default_library",
+        "@org_golang_google_grpc//:go_default_library",
+        "@org_golang_google_grpc//codes:go_default_library",
+        "@org_golang_google_grpc//grpclog:go_default_library",
+        "@org_golang_google_grpc//status:go_default_library",
+        "@org_golang_x_net//context:go_default_library",
+    ],
```
Comment (jayconrod) [inline proto/BUILD.bazel]: What about other Well Known Types?

File: proto/grpc_gateway.bzl [hunk 4]
```diff
@@ -0,0 +1,40 @@
+GOOGLEAPIS_GOOGLE_API_BUILD_CONTENTS = """
```
Comment (jayconrod) [inline proto/grpc_gateway.bzl]: Echoing the comment above, this should be possible with `http_archive` with an `overlay`. I don't think this file will be needed.

File: third_party/com_github_grpc_ecosystem_grpc_gateway/WORKSPACE [hunk 5]
```diff
@@ -0,0 +1 @@
+workspace(name = "com_github_grpc_ecosystem_grpc_gateway")
```
Comment (jayconrod) [inline third_party/com_github_grpc_ecosystem_grpc_gateway/WORKSPACE]: You might be able to delete this file. I think Bazel creates it automatically in external repositories.

File: third_party/com_github_grpc_ecosystem_grpc_gateway/protoc-gen-grpc-gateway/BUILD.in [hunk 6]
```diff
@@ -0,0 +1,41 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library")
```
Comment (jayconrod) [inline third_party/com_github_grpc_ecosystem_grpc_gateway/protoc-gen-grpc-gateway/BUILD.in]: This repo already has build files. Why do they need to be replaced? Can they be modified there instead?

**Comment by achew22** [comment]
@Mistobaan, QQ what is the benefit of pushing these up into the `rules_go` repo when the compiler is already defined in [grpc_gateway](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/protoc-gen-grpc-gateway/BUILD.bazel#L25)?

File: third_party/com_github_grpc_ecosystem_grpc_gateway/protoc-gen-grpc-gateway/BUILD.in (hunk 6 above)
Comment (Mistobaan) [reply third_party/com_github_grpc_ecosystem_grpc_gateway/protoc-gen-grpc-gateway/BUILD.in]: yes that is the best strategy. see comments below.

File: proto/grpc_gateway.bzl (hunk 4 above)
Comment (Mistobaan) [reply proto/grpc_gateway.bzl]: but still defined within rules_go in the repository.bzl right?

**Comment by Mistobaan** [comment]
> I'm not directly familiar with grpc-gateway. If issues are reported on these rules, will you be able to help support them for a while?

Yes 

> It sounds like some changes are needed upstream in grpc-gateway. Those should land first before we proceed. The Gazelle changes can happen after this.

> CI tests are failing due to missing dependencies. What other dependencies are required? You can run bazel test //... locally before pushing. Our CI doesn't do much else.

yes grpc_gateway has some dependencies declared in its own WORKSPACE. I am not sure what is the best way to include these dependencies inside rules_go beside also copying them directly into repositories.bzl. Any suggestions?

> We need some additional tests to verify that builds with these rules are working. Please add something under tests/core/grpc_gateway.

Will do

**Comment by jayconrod** [comment]
@achew22 @Mistobaan 

High level question: does it make sense, architecturally, for rules_go to provide grpc-gateway support? Is there a problem that can't be solved another way?

My concerns are:
* There are a lot of dependencies. These shouldn't be fetched most of the time, but they're defaulting people to specific versions which may be unexpected. People will be annoyed when these are updated (or aren't updated).
* I'm not familiar enough with grpc-gateway to support or maintain rules for it. If rules were defined and actively maintained in grpc-gateway, I think that would be a better outcome for users.
* In general, I think rules_go should have as few dependencies as possible. It makes sense to provide `go_proto_library`, `go_proto_compiler`, WKTs, and a small set of plugins (`go_proto`, `go_grpc`), I don't think we should try to provide all plugins.

I *do* think it makes sense for Gazelle to support these. If our problems can be solved by Gazelle setting `go_proto_library` `compiler ` attributes and adding rules in WORKSPACE as needed, I think that's a better solution. Gazelle can go beyond what rules_go supports.

cc @ashi009 who also has a pending PR adding grpc-gateway support that I have unfortunately neglected for several months (sorry sorry sorry)

**Comment by Mistobaan** [comment]
> In general, I think rules_go should have as few dependencies as possible. It makes sense to provide go_proto_library, go_proto_compiler, WKTs, and a small set of plugins (go_proto, go_grpc), I don't think we should try to provide all plugins.

I agree. The fact that gogo was included here and the other PRs mislead me. 
Then we could think how to add to Gazelle some kind of proto_compiler_plugin. 

on a side note, I do think the googleapis rules do belongs in rules_go, so will change this PR to just include the googleapis (or make a new issue + PR if you prefer).

**Comment by jayconrod** [comment]
Closing this PR per earlier discussion. I think having grpc-gateway provide its own `go_proto_compiler` makes the most sense, architecturally. Ideally, rules_go should only support the default plugin and gRPC.

Agree this is inconsistent with what we do for gogo. Ideally, gogo support would not live here.

**Pull request closed without merging**